businessLogicLayer/Features/PricesNearMe/StationDataManager.swift

```diff
@@ -0,0 +1,123 @@
+//  fuelPrices
+//
+//  Created by Robson Harrison on 25/04/2024.
+//
+
+import Foundation
+
+// The StationDataManager is the API (or interface) to this feature of the app. The caller of this class should not need any other knowledge of any internal workings of HOW this class achieves its purpose.
+class StationDataManager: ObservableObject {
+    
+    enum StationDataManagerError: Error {
+        case stationDataAPIError(StationDataAPI.StationDataAPIError)
+        case refreshAlreadyInProgress
+    }
+    
+    let fuelProviders: [String: String]
+    private var dateOfLastRefresh: Date?
+    private(set) var isRefreshingData = false
+    // No layer above this class should know that StationDataAPI exists as a service.
+    private var stationDataAPI = StationDataAPI()
+    private var stationLibrary: [PetrolStation] = []
+    
+    // MARK: - Constructor
+    
+    init(fuelProviders: [String: String]) {
+        // a constructor (initialiser) should only initialize its properties. It shouldn't be used to trigger network calls or any other hidden behaviour.
+        self.fuelProviders = fuelProviders
+    }
+    
+    // MARK: - Publicly available functions / features (make it easy-to-use)
+    
+    func getStationData(forPostcode postcode: String, completion: @escaping (Result<[PetrolStation], StationDataManagerError>) -> Void) {
+        guard isRefreshingData == false else {
+            // Note: ALWAYS call your completion handlers in 100% of scenarios 😉
+            completion(.failure(.refreshAlreadyInProgress))
+            return
+        }
+        isRefreshingData = true
+        let endpointAddresses = fuelProviders.map{$1}
+        gatherStationData(from: endpointAddresses) { result in
+            DispatchQueue.main.async {
+                self.isRefreshingData = false
+                switch result {
+                case .success(let stations):
+                    self.dateOfLastRefresh = Date.now
+                    // now that we are running back on the main thread we can set the array of stations to this instance of StationDataManager. But not before. Unless we use a queue to handle read and write access.
+                    self.stationLibrary = stations // update the library of national petrol stations
+                    let filteredResults = self.filterStations(stations, forPostcode: postcode)
+                    completion(.success(filteredResults))
+                case .failure(let error):
+                    completion(result)
+                }
+            }
+        }
+    }
+    
+    // MARK: - Gathering data from StationDataAPI
+    
+    func gatherStationData(from endPointAddresses: [String], completion: @escaping (Result<[PetrolStation], StationDataManagerError>) -> Void) {
+        let stationDataAPI = self.stationDataAPI
+        let dispatchGroup = DispatchGroup()
+        var responses: [ResponseData] = []
+        
+        for (endPointAddress) in endPointAddresses {
```

Comment:
So, this `StationDataManager` class now handles looping through each url and manipulates the data returned. This is cool because it keeps the API class true to its purpose which is to make one network call to a specific end point (url).  The `StationDataManager` is the layer above, which is now deciding to use the API in a certain way but doesn't affect how it's designed because of its own needs. This would allow the API class to be used again by anything that just wants to request petrol station data. 

I hope that made some sense. :)